teachertool/src/components/DragAndDropFileSurface.tsx

```diff
@@ -0,0 +1,103 @@
+import { classList } from "react-common/components/util";
+import { Strings } from "../constants";
+import { NoticeLabel } from "./NoticeLabel";
+import { useRef, useState } from "react";
+import css from "./styling/DragAndDropFileSurface.module.scss";
+import { Button } from "react-common/components/controls/Button";
+
+export interface DragAndDropFileSurfaceProps {
+    onFileDroppedAsync: (file: File) => void;
+    errorMessage?: string;
+}
+export const DragAndDropFileSurface: React.FC<DragAndDropFileSurfaceProps> = ({ onFileDroppedAsync, errorMessage }) => {
+    const [fileIsOverSurface, setFileIsOverSurface] = useState(false);
+    const [errorKey, setErrorKey] = useState(0); // Used to reset animations on the error label.
+    const fileInputRef = useRef<HTMLInputElement>(null);
+
+    function handleDragOver(event: React.DragEvent<HTMLDivElement>) {
+        // Stop the browser from intercepting the file.
+        event.preventDefault();
+    }
+
+    function handleDragEnter(event: React.DragEvent<HTMLDivElement>) {
+        event.preventDefault();
+        setFileIsOverSurface(true);
+    }
+
+    function handleDragLeave(event: React.DragEvent<HTMLDivElement>) {
+        event.preventDefault();
+        setFileIsOverSurface(false);
+    }
+
+    function handleDrop(event: React.DragEvent<HTMLDivElement>) {
+        event.preventDefault();
+
+        setFileIsOverSurface(false);
+
+        const file = event.dataTransfer.files[0];
+        if (file) {
+            processNewFile(file);
+        }
+    }
+
+    function handleFileFromBrowse(event: React.ChangeEvent<HTMLInputElement>) {
+        const file = event.target.files?.[0];
+        if (file) {
+            processNewFile(file);
+        }
+    }
+
+    function processNewFile(file: File) {
+        setErrorKey(errorKey + 1);
+        onFileDroppedAsync(file);
+    }
+
+    /*
+    We can't use the drag-and-drop-file-surface directly to handle most drop events, because the child elements interfere with them.
+    To solve this, we add a transparent div (droppable-surface) over everything and use that for most drag-related event handling.
+    However, we don't want the transparent droppable-surface to intercept pointer events when there is no drag occurring, so
+    we still use the drag-and-drop-file-surface to detect dragEnter events and only intercept pointer events on the droppable-surface
+    after that has happened.
+    */
+    return (
+        <div className={css["drag-and-drop-file-surface"]} onDragEnter={handleDragEnter}>
+            <div className={css["instruction-container"]}>
+                <i className={classList("fas fa-file-upload", css["upload-icon"])}></i>
+                <div className="no-select">{fileIsOverSurface ? Strings.ReleaseToUpload : Strings.DragAndDrop}</div>
```

Comment:
Do we need to wrap `no-select` in `css["no-select"]`?

Reply:
Same for the "hidden" classname for the input below.

Reply:
Nope! These classes come from react-common which means the names haven't been modified by the scss stuff, so no lookup is necessary.